icu4c/source/tools/upropdump/upropdump.cpp

```diff
@@ -0,0 +1,178 @@
+        status));
+    handleError(status, fullPropName);
+
+    fputs("[code_point_trie.struct]\n", f);
+    fprintf(f, "long_name = \"%s\"\n", fullPropName);
+    usrc_writeUCPTrie(f, shortPropName, utrie.getAlias(), UPRV_TARGET_SYNTAX_TOML);
+}
+
+enum {
+    OPT_HELP_H,
+    OPT_HELP_QUESTION_MARK,
+    OPT_COPYRIGHT,
+    OPT_VERSION,
+    OPT_DESTDIR,
+    OPT_VERBOSE,
+    OPT_QUIET,
+
+    OPT_COUNT
+};
+
+static UOption options[]={
+    UOPTION_HELP_H,
+    UOPTION_HELP_QUESTION_MARK,
+    UOPTION_COPYRIGHT,
+    UOPTION_VERSION,
+    UOPTION_DESTDIR,
+    UOPTION_VERBOSE,
+    UOPTION_QUIET,
+};
+
+int main(int argc, char* argv[]) {
+
+    U_MAIN_INIT_ARGS(argc, argv);
+
+    /* preset then read command line options */
+    options[OPT_DESTDIR].value=u_getDataDirectory();
+    argc=u_parseArgs(argc, argv, UPRV_LENGTHOF(options), options);
+
+    if(options[OPT_VERSION].doesOccur) {
+        printf("upropdump version %s, ICU tool to write Unicode property .toml files\n",
+               U_ICU_DATA_VERSION);
+        printf("%s\n", U_COPYRIGHT_STRING);
+        exit(0);
+    }
+
+    /* error handling, printing usage message */
+    if(argc<0) {
+        fprintf(stderr,
+            "error in command line argument \"%s\"\n",
+            argv[-argc]);
+    } else if(argc<2) {
+        argc=-1;
+    }
+
+    if(argc<0 || options[OPT_HELP_H].doesOccur || options[OPT_HELP_QUESTION_MARK].doesOccur) {
+        FILE *stdfile=argc<0 ? stderr : stdout;
+        fprintf(stdfile,
+            "usage: %s [-options] properties...\n"
+            "\tdump Unicode property data to .toml files\n"
+            "options:\n"
+            "\t-h or -? or --help  this usage text\n"
+            "\t-V or --version     show a version message\n"
+            "\t-c or --copyright   include a copyright notice\n"
+            "\t-d or --destdir     destination directory, followed by the path\n"
+            "\t-v or --verbose     Turn on verbose output\n"
+            "\t-q or --quiet       do not display warnings and progress\n",
+            argv[0]);
+        return argc<0 ? U_ILLEGAL_ARGUMENT_ERROR : U_ZERO_ERROR;
+    }
+
+    /* get the options values */
+    haveCopyright = options[OPT_COPYRIGHT].doesOccur;
+    const char *destdir = options[OPT_DESTDIR].value;
+    VERBOSE = options[OPT_VERBOSE].doesOccur;
+    QUIET = options[OPT_QUIET].doesOccur;
+
+    for (int i=1; i<argc; i++) {
+        auto propName = argv[i];
+        UProperty propEnum = u_getPropertyEnum(propName);
+        if (propEnum == UCHAR_INVALID_CODE) {
+            std::cerr << "Error: Invalid property alias: " << propName << std::endl;
+            exit(1);
+        }
+
+        ErrorCode status;
+        CharString outFileName;
+        if (destdir != nullptr || *destdir != 0) {
+            outFileName.append(destdir, status).ensureEndsWithFileSeparator(status);
+        }
+        outFileName.append(propName, status);
```

Comment:
Please "canonicalize" the propName for the purpose of creating the output file --> `u_getPropertyName(propEnum, U_LONG_PROPERTY_NAME)`

For the file name, I would use the long name because it's more readable (and more findable). The short ones can be *very* obscure.

Note that each of your dumpXyzProperty() functions already get the standard name strings.
Simplest might be to pull more of their code up to here and share it -- getting & printing names and the `[xyz.data]` line.
You could inline dumpBinaryProperty() completely.
You would still call a reduced dumpEnumeratedProperty() for getting/building/writing the map & trie.

Reply:
I considered what to use as the filename (input, short canonical, or long canonical?).  I decided to use the input because then the caller can control the filename.  If they want short canonical filenames, then they can pass short canonical command line arguments, and if they want long canonical filenames, then they can pass long canonical command line arguments.